feat: Introduce GitHub Action

nderjung opened the pull request:
<!--

Thank you for opening a new PR to the Unikraft Open Source Project!  We welcome
new changes, features, fixes, and more!  Please fill in this form to indicate
the status of your PR.  Please ensure you have read the contribution guidelines
before opening a new PR as this will cover the PR process:

  https://unikraft.org/docs/contributing/

  Kraftkit follows the same guidelines as the Unikraft Open Source Project.

-->

### Prerequisite checklist

<!--
Please mark items appropriately:
-->

  - [x] Read the [contribution guidelines](https://unikraft.org/docs/contributing/) regarding submitting new changes to the project;
  - [x] Tested your changes against relevant architectures and platforms;
  - [x] Ran `make fmt` on your commit series before opening this PR;
  - [ ] Updated relevant documentation.

### Description of changes

<!--
Please provide a detailed description of the changes made in this new PR.
-->

This PR introduces a new tool, `github-action`, and a definition file that turns the repository into a GitHub Action such that it can be used in workflows via the following step syntax:

```yaml
- uses: unikraft/kraftkit@stable
  with:
    workdir: path/to/project
    kraftfile: path/to/Kraftfile
    arch: x86_64
    plat: qemu
    push: false
    output: oci://ghcr.io/username/repo/unikernel:latest
    kconfig: true
```

This PR facilitates the introduction of a new GitHub Action step and has made the following additional adjustments to make this possible:
- Upgraded QEMU to v7.2.4, representing the latest version;
- Fixed an issue where QEMU's TCG accelerator was not used on a KVM-enabled host even though it was requested; this is due to misconfiguration during the build and force enabling system and user emulation;
- Ensuring that if GCC, QEMU builds are affected, the GitHub action's image is also re-built, ensuring a stable result;
- Add libc6-dev dependency to enable GCC builds for arm{64}; and,
- Recognising the `Output` attribute in the OCI PM handler.

### TODO

- [x] Define container image in `buildenvs/github-action.Dockerfile`
- [x] New GitHub actions workflow to build the tool and the container image;
- [x] "Run" option to run the image in emulation mode;
- [x] New "Integrations" workflow that actually runs the github action with a simple helloworld app (build then run);
- [x] Push artifact based on output name, i.e. if output is `oci://` use OCI PM.

craciunoiuc committed:
feat(tools): Introduce github-action tool

This tool is designed to be used in a CI/CD context on GitHub in an
actions workflow and aims to simplify the entire process of using
KraftKit by offering several high-level abstractions that are useful
in this context.

The ultimate use case is for this new GitHub Action is when a user
has a Kraftfile in their repository and to get started building,
packaging and publish the image by simply introducing a small, well-
defined uses step. The following is an example `uses` step which
this tool ultimately aims to offer:

```yaml
- uses: unikraft/kraftkit@stable
  with:
    workdir: path/to/project
    kraftfile: path/to/Kraftfile
    arch: x86_64
    plat: qemu
    push: false
    output: oci://ghcr.io/username/repo/unikernel:latest
    kconfig: true
```

Signed-off-by: Cezar Craciunoiu <cezar.craciunoiu@unikraft.io>
Co-authored-by: Alexander Jung <alex@unikraft.io>

craciunoiuc committed:
feat(buildenvs): Add github-action tool build definition

Signed-off-by: Cezar Craciunoiu <cezar.craciunoiu@unikraft.io>
Co-authored-by: Alexander Jung <alex@unikraft.io>

craciunoiuc committed:
feat(Makefile): Register github-action container build & group tools

Signed-off-by: Cezar Craciunoiu <cezar.craciunoiu@unikraft.io>

craciunoiuc committed:
feat(workflows): Add workflow to build github-action container tool

Signed-off-by: Cezar Craciunoiu <cezar.craciunoiu@unikraft.io>

craciunoiuc committed:
feat(action): Register action to be usable with github-action

Signed-off-by: Cezar Craciunoiu <cezar.craciunoiu@unikraft.io>

antoineco commented:
Could you please explain how this differs from running `kraft` commands in a generic GitHub Action step? The description explains _WHAT_ but not _WHY_.

nderjung commented:
Hi @antoineco. Of course if a user wishes to simply run a container image and call `kraft` under neath, this is perfectly possible. This tool aims to simplify the entire process by only offering several high-level abstractions that are useful in a CI/CD-context.

The ultimate use case is for this new GitHub Action is when a user has a `Kraftfile` in their repository and to get started building, packaging and publish the image by simply introducing a small, well-defined `uses` step. The example step in the PR message demonstrates a very simple API for building a project.

One additional reason for providing such a high-level utility, in my opinion and which I worry about, is that `kraft`'s CLI flags are still considered unstable. Recently, `kraft pkg --oci-tag` was renamed with `kraft pkg --name`. Towards stabilising the use of Kraft_Kit_, this new actions step will be provided as the recommended method to build unikernels programmatically on GitHub in pipelines without having to worry about the `kraft` binary underneath. Flag changes could suddenly render many external pipelines erroring.

Beyond this use case, I imagine a more long-term goal is to support the wider Unikraft community with this action with additional options that allow for parsing the results from the stdout of a unikernel. This is not yet possible, nor in the core, but [is planned via a modification to the internal `ukprint` library](https://github.com/unikraft/unikraft/issues/665) which will allow for outputting results in JSON format. Post-processing steps can then be done to render high-level information in the form of PR comments (similar to how @unikraft-bot outputs a checkpatch report).  One use case here is running Unikraft's internal testsuite, `uktest`, and rendering glance-level information about which tests failed and why. This is can then be used by anyone but also by all official Unikraft `lib-*` repositories.

nderjung committed:
feat: Upgrade QEMU from v7.1.0 to v7.2.4

This is the latest version of the v7 series.  At the same time, set the
QMP version such that it is reflected via the import.  We are setting
`v7alpha2` here instead of `v7` because the QMP implementation is
incomplete.  If it is fully populated we can rename to `v7`,
representing a complete API that can then be referenced.

Signed-off-by: Alexander Jung <alex@unikraft.io>

nderjung committed:
fix(buildenvs): Force setting system and user mode in QEMU build

This fix adjusts the build of the QEMU programs and force enables system
emulation and user emulation.  Without these options set, TCG
acceleration cannot be done and the `qemu-system-*` binary always
searches for host acceleration.

Signed-off-by: Alexander Jung <alex@unikraft.io>

nderjung committed:
feat(buildenvs): Enable aarch64 and arm in QEMU build

Signed-off-by: Alexander Jung <alex@unikraft.io>

nderjung committed:
fix(buildenvs): Add libc6-dev and sort dependencies alphabetically

Signed-off-by: Alexander Jung <alex@unikraft.io>

nderjung committed:
fix(oci): Adjust and allow for using full output or partial name

If a full "output" is provided, use this entirely, otherwise use the
default registry along with the supplied name, falling back further to
the target name.

Signed-off-by: Alexander Jung <alex@unikraft.io>

nderjung commented:
@antoineco this is ready for a pass :)
